Stripe/STPPostalCodeValidator.m

```diff
@@ -7,40 +7,171 @@
 //
 
 #import "STPPostalCodeValidator.h"
+
 #import "STPCardValidator.h"
 #import "STPPhoneNumberValidator.h"
+#import "NSCharacterSet+Stripe.h"
+#import "NSCharacterSet+Stripe.h"
+#import "NSString+Stripe.h"
+
+static NSString *const STPCountryCodeUnitedStates = @"US";
 
 @implementation STPPostalCodeValidator
 
-+ (BOOL)stringIsValidPostalCode:(nullable NSString *)string
-                           type:(STPPostalCodeType)postalCodeType {
-    switch (postalCodeType) {
-        case STPCountryPostalCodeTypeNumericOnly: {
-            NSInteger length = [STPCardValidator sanitizedNumericStringForString:string].length;
-            return (length == 5 || length == 9);
++ (STPCardValidationState)validationStateForPostalCode:(NSString *)postalCode
+                                           countryCode:(NSString *)countryCode {
+    NSString *sanitizedCountryCode = countryCode.uppercaseString;
+    if ([self postalCodeIsRequiredForCountryCode:countryCode]) {
+        if ([sanitizedCountryCode isEqualToString:STPCountryCodeUnitedStates]) {
+            return [self validationStateForUSPostalCode:postalCode];
+        }
+        else {
+            if (postalCode.length > 0) {
+                return STPCardValidationStateValid;
+            }
+            else {
+                return STPCardValidationStateIncomplete;
+            }
         }
-        case STPCountryPostalCodeTypeAlphanumeric:
-            return string.length > 0;
-        case STPCountryPostalCodeTypeNotRequired:
-            return YES;
+    }
+    else {
+        return STPCardValidationStateValid;
     }
 }
 
-+ (BOOL)stringIsValidPostalCode:(nullable NSString *)string
-                    countryCode:(nullable NSString *)countryCode {
-    return [self stringIsValidPostalCode:string
-                                    type:[self postalCodeTypeForCountryCode:countryCode]];
+static NSUInteger countOfCharactersFromSetInString(NSString * _Nonnull string, NSCharacterSet * _Nonnull cs) {
+    NSRange range = [string rangeOfCharacterFromSet:cs];
+    NSUInteger count = 0;
+    if (range.location != NSNotFound) {
+        NSUInteger lastPosition = NSMaxRange(range);
+        count += range.length;
+        while (lastPosition < string.length) {
+            range = [string rangeOfCharacterFromSet:cs options:(NSStringCompareOptions)kNilOptions range:NSMakeRange(lastPosition, string.length - lastPosition)];
+            if (range.location == NSNotFound) {
+                break;
+            }
+            else {
+                count += range.length;
+                lastPosition = NSMaxRange(range);
+            }
+        }
+    }
+
+    return count;
```

Comment:
this function is a little hairy but I'm guessing it's the right way to handle the various string encodings?